data/transactions/logic/eval_test.go

```diff
@@ -5117,3 +5117,291 @@ func TestOpBase64Decode(t *testing.T) {
+			byte "key2"; 
+			json_ref JSONObject; 
+			byte "key3";
+			json_ref JSONString;
+			byte ""teal"" // quotes match
+			==`,
+		},
+		{
+			source: `byte  "{\"key0\": 0,\"key1\": \"algo\",\"key2\":{\"key3\": \" teal \", \"key4\": {\"key40\": 10}}, \"key5\": 18446744073709551615 }"; 
+			byte "key2"; 
+			json_ref JSONObject; 
+			byte "key3";
+			json_ref JSONString;
+			byte " teal " // spaces match
+			==`,
+		},
+		{
+			source: `byte  "{\"key0\": 0,\"key1\": \"algo\",\"key2\":{\"key3\": \"teal\", \"key4\": {\"key40\": 10, \"key40\": \"10\"}}, \"key5\": 18446744073709551615 }"; 
+			byte "key2"; 
+			json_ref JSONObject; 
+			byte "key4";
+			json_ref JSONObject;
+			byte "{\"key40\": 10, \"key40\": \"10\"}"
+			==
+			`,
+		},
+		{
+			source: `byte  "{\"rawId\": \"responseId\",\"id\": \"0\",\"response\": {\"attestationObject\": \"based64url_encoded_buffer\",\"clientDataJSON\":  \" based64url_encoded_client_data\"},\"getClientExtensionResults\": {},\"type\": \"public-key\"}"; 
+			byte "response"; 
+			json_ref JSONObject; 
+			byte "{\"attestationObject\": \"based64url_encoded_buffer\",\"clientDataJSON\":  \" based64url_encoded_client_data\"}" // object as it appeared in input
+			==`,
+		},
+		{
+			source: `byte  "{\"rawId\": \"responseId\",\"id\": \"0\",\"response\": {\"attestationObject\": \"based64url_encoded_buffer\",\"clientD\\u0061taJSON\":  \" based64url_encoded_client_data\"},\"getClientExtensionResults\": {},\"type\": \"public-key\"}"; 
+			byte "response"; 
+			json_ref JSONObject; 
+			byte "{\"attestationObject\": \"based64url_encoded_buffer\",\"clientD\\u0061taJSON\":  \" based64url_encoded_client_data\"}" // object as it appeared in input
+			==`,
+		},
+		{
+			source: `byte  "{\"rawId\": \"responseId\",\"id\": \"0\",\"response\": {\"attestationObject\": \"based64url_encoded_buffer\",\"clientDataJSON\":  \" based64url_encoded_client_data\"},\"getClientExtensionResults\": {},\"type\": \"public-key\"}"; 
+			byte "response"; 
+			json_ref JSONObject; 
+			byte "clientDataJSON"; 
+			json_ref JSONString; 
+			byte " based64url_encoded_client_data"; 
+			==`,
+		},
+		{
+			source: `byte  "{\"\\u0072\\u0061\\u0077\\u0049\\u0044\": \"responseId\",\"id\": \"0\",\"response\": {\"attestationObject\": \"based64url_encoded_buffer\",\"clientDataJSON\":  \" based64url_encoded_client_data\"},\"getClientExtensionResults\": {},\"type\": \"public-key\"}"; 
+			byte "rawID"; 
+			json_ref JSONString; 
+			byte "responseId"
+			==`,
+		},
+	}
+
+	for _, s := range testCases {
+		ops := testProg(t, s.source, AssemblerMaxVersion)
+
+		err := CheckStateful(ops.Program, ep)
+		require.NoError(t, err, s)
+
+		pass, _, err := EvalStatefulCx(ops.Program, ep)
+		require.NoError(t, err)
+		require.True(t, pass)
+	}
+
+	failedCases := []struct {
+		source string
+		error  string
+	}{
+		{
+			source: `byte  "{\"key0\": 1 }"; byte "key0"; json_ref JSONString;`,
```

Comment:
I don't think you have failing test cases for the following yet?
* Calling `json_ref` on a JSON array, e.g. `byte "[1,2,3]"; byte "key"; json_ref JSONUint64`
* Calling `json_ref` on a JSON integer, e.g. `byte "2"; byte "key"; json_ref JSONUint64`
* Calling `json_ref` on a JSON null, e.g. `byte "null"; byte "key"; json_ref JSONUint64`
* Calling `json_ref` on a JSON string, e.g. `byte "\"string\""; byte "key"; json_ref JSONString`

I'm pretty sure they'll fail, but since `hasDuplicateKeys` does some interesting things with the tokenizer, I'm worried it might panic instead of failing gracefully

Reply:
good point. tests added.